waspc/src/Wasp/Generator/PackageJsonGenerator.hs

```diff
@@ -1,93 +1,124 @@
+
+data PackageJsonDependenciesError = PackageJsonDependenciesError
+  { dependenciesConflictErrors :: [DependencyConflictError],
+    devDependenciesConflictErrors :: [DependencyConflictError]
+  }
+  deriving (Show, Eq)
+
+data DependencyConflictError = DependencyConflictError
+  { waspDependency :: D.Dependency,
+    userDependency :: D.Dependency
+  }
   deriving (Show, Eq)
 
 -- | Takes wasp npm dependencies and user npm dependencies and figures out how
---   to combine them together, returning (Right) a list of npm dependencies to
---   be used on behalf of wasp and then also a list of npm dependencies to be
---   used on behalf of user. These lists might be the same as the initial ones,
---   but might also be different.
+--   to combine them together, returning (Right) a new PackageJsonDependencies
+--   that combines them, and on error (Left), returns a PackageJsonDependenciesError
+--   which describes which dependencies are in conflict.
+combinePackageJsonDependencies :: PackageJsonDependencies -> PackageJsonDependencies -> Either PackageJsonDependenciesError PackageJsonDependencies
+combinePackageJsonDependencies waspDependencies userDependencies =
+  if null conflictErrors && null devConflictErrors
+    then
+      Right $
+        PackageJsonDependencies
+          { dependencies = dependencies waspDependencies ++ remainingUserDeps,
+            devDependencies = devDependencies waspDependencies ++ remainingUserDevDeps
+          }
+    else
+      Left $
+        PackageJsonDependenciesError
+          { dependenciesConflictErrors = conflictErrors,
+            devDependenciesConflictErrors = devConflictErrors
+          }
+  where
+    waspDepsByName = makeDepsByName $ dependencies waspDependencies
+    waspDevDepsByName = makeDepsByName $ devDependencies waspDependencies
+    userDepsByName = makeDepsByName $ dependencies userDependencies
+    userDevDepsByName = makeDepsByName $ devDependencies userDependencies
+    allWaspDepsByName = waspDepsByName `Map.union` waspDevDepsByName
+    conflictErrors = getConflictErrors allWaspDepsByName userDepsByName
+    devConflictErrors = getConflictErrors allWaspDepsByName userDevDepsByName
+    remainingUserDeps = getRemainingUserDeps waspDepsByName userDepsByName
+    remainingUserDevDeps = getRemainingUserDeps waspDevDepsByName userDevDepsByName
 
---   On error (Left), returns a list of conflicting wasp dep, user dep pairs.
-resolveDependencies ::
-  [D.Dependency] ->
-  [D.Dependency] ->
-  Either
-    [DependencyConflictError]
-    ([D.Dependency], [D.Dependency])
-resolveDependencies waspDeps userDeps =
-  if null conflictErrors
-    then Right (waspDeps, Map.elems userDepsNotInWaspDeps)
-    else Left conflictErrors
+-- Given a map of wasp dependencies and a map of user dependencies, construct a
+-- list of conflict errors for any dependencies that have a matching name but
+-- different version.
+getConflictErrors :: DepsByName -> DepsByName -> [DependencyConflictError]
```

Comment:
I've decoupled conflict generation from removing overlaps from the userDeps. This way I can do conflict detection on allWaspDepsByName, whereas getting the overlaps can be done with either dev or non-dev dependencies.